TrackletAlgorithm/TrackMerger.cc

```diff
@@ -6,167 +6,162 @@ TrackStruct& ComparisonModule::getMasterTrackStruct(){
 ) {
-  trackWord_o.write(aTrack._trackWord);
-  barrelStubWords_0_o.write(aTrack._barrelStubArray[0][0]);
-  barrelStubWords_1_o.write(aTrack._barrelStubArray[1][0]);
-  barrelStubWords_2_o.write(aTrack._barrelStubArray[2][0]);
-  barrelStubWords_3_o.write(aTrack._barrelStubArray[3][0]);
-  diskStubWords_0_o.write(aTrack._diskStubArray[0][0]);
-  diskStubWords_1_o.write(aTrack._diskStubArray[1][0]);
-  diskStubWords_2_o.write(aTrack._diskStubArray[2][0]);
-  diskStubWords_3_o.write(aTrack._diskStubArray[3][0]);
+
+  #pragma HLS array_partition variable=barrelStubWordsOut
+  #pragma HLS array_partition variable=diskStubWordsOut
+
+  trackWordOut = aTrack.getTrkWord(); 
+  for (unsigned int i = 0; i < NBarrelStub; i++){
+    #pragma HLS unroll
+    barrelStubWordsOut[i] = aTrack.getBarrelStub(i, 0);
+    }
+
+  for (unsigned int j = 0; j < NDiskStub; j++){
+    #pragma HLS unroll
+    diskStubWordsOut[j]= aTrack.getDiskStub(j, 0);
+  }
 
   return;
 }
 
 void TrackMerger(const BXType bx,
-  stream<TrackFitType::TrackWord> &trackWord,
-  stream<TrackFitType::BarrelStubWord> &barrelStubWords_0,
-  stream<TrackFitType::BarrelStubWord> &barrelStubWords_1,
-  stream<TrackFitType::BarrelStubWord> &barrelStubWords_2,
-  stream<TrackFitType::BarrelStubWord> &barrelStubWords_3,
-  stream<TrackFitType::DiskStubWord> &diskStubWords_0,
-  stream<TrackFitType::DiskStubWord> &diskStubWords_1,
-  stream<TrackFitType::DiskStubWord> &diskStubWords_2,
-  stream<TrackFitType::DiskStubWord> &diskStubWords_3,
+  const TrackFitType::TrackWord trackWord[kMaxTrack],
+  const TrackFitType::BarrelStubWord barrelStubWords[kMaxTrack][NBarrelStub],
+  const TrackFitType::DiskStubWord diskStubWords[kMaxTrack][NDiskStub],
   BXType bx_o,
-  stream<TrackFitType::TrackWord> &trackWord_o,
-  stream<TrackFitType::BarrelStubWord> &barrelStubWords_0_o, 
-  stream<TrackFitType::BarrelStubWord> &barrelStubWords_1_o, 
-  stream<TrackFitType::BarrelStubWord> &barrelStubWords_2_o, 
-  stream<TrackFitType::BarrelStubWord> &barrelStubWords_3_o, 
-  stream<TrackFitType::DiskStubWord> &diskStubWords_0_o, 
-  stream<TrackFitType::DiskStubWord> &diskStubWords_1_o, 
-  stream<TrackFitType::DiskStubWord> &diskStubWords_2_o, 
-  stream<TrackFitType::DiskStubWord> &diskStubWords_3_o 
+  TrackFitType::TrackWord (&trackWord_o)[kMaxTrack],
+  TrackFitType::BarrelStubWord (&barrelStubWords_o)[kMaxTrack][NBarrelStub],
+  TrackFitType::DiskStubWord (&diskStubWords_o)[kMaxTrack][NDiskStub]
   )
 {
   static ComparisonModule comparisonModule[kNComparisonModules];
-  #pragma HLS array_partition variable=comparisonModule complete dim=0
-
-  // TrackStruct unmergedTracks[kMaxTrack];
-  // #pragma HLS array_partition variable=unmergedTracks complete dim=0
+  #pragma HLS array_partition variable=comparisonModule dim=1
+
+  static TrackStruct unmergedTracks[kNLastTracks];
+  #pragma HLS data_pack variable=unmergedTracks
+
+  static TrackStruct cmTracks[kNComparisonModules];
```

Comment:
@kwalkingshaw Do these actually need to be `static`? Does removing the `static` keyword change the results of the C-simulation or C-synthesis?

Reply:
Removing the static doesn't update the arrays properly (even in simulation) and the static keyword also reduces the latency as extra loops appeared in the synthesis report which had a latency of the size of the array, not sure if this is an HLS issue?

Reply:
The constructor for `ComparisonModule` initializes members of the class:
https://github.com/cms-L1TK/firmware-hls/blob/e794563ac2d4289c9471265ecdfebb974b87cd93/TrackletAlgorithm/TrackMerger.h#L17-L18

And you also have default member initializations in both `ComparisonModule` (which obviate the initializations in the constructor, BTW) and `TrackStruct`:
https://github.com/cms-L1TK/firmware-hls/blob/e794563ac2d4289c9471265ecdfebb974b87cd93/TrackletAlgorithm/TrackMerger.h#L37-L41
https://github.com/cms-L1TK/firmware-hls/blob/e794563ac2d4289c9471265ecdfebb974b87cd93/TrackletAlgorithm/TrackHandler.h#L22-L24

When declaring an array of `ComparisonModule` or `TrackStruct` objects, the compiler will generate an implicit loop to initialize all these members. Unfortunately, being implicit means you cannot unroll the loop as you normally would with the HLS unroll pragma.

So you could try moving all these initializations to separate methods (assuming they are all truly needed), and then calling these methods in explicit loops after declaring your arrays of `ComparisonModule`/`TrackStruct`. Then you can just add the unroll pragma to these loops, e.g., as is done in the TrackletProcessor:
https://github.com/cms-L1TK/firmware-hls/blob/e794563ac2d4289c9471265ecdfebb974b87cd93/TrackletAlgorithm/TrackletProcessor.h#L569-L575

However, I'm more confused that the arrays don't update properly without `static`. Does that mean the results of one event depend on something happening in the previous event? Or am I missing something?